engine/consensus/approvals/assignment_collector_tree.go

```diff
@@ -12,20 +12,40 @@ import (
 // assignmentCollectorVertex is a helper structure that implements a LevelledForrest Vertex interface and encapsulates
 // AssignmentCollector and information if collector is processable or not
 type assignmentCollectorVertex struct {
-	collector   *AssignmentCollector
-	processable bool
+	collector               *VerifyingAssignmentCollector
+	nonProcessableCollector AssignmentCollector
+	processable             bool
 }
 
 /* Methods implementing LevelledForest's Vertex interface */
 
-func (v *assignmentCollectorVertex) VertexID() flow.Identifier { return v.collector.ResultID }
+func (v *assignmentCollectorVertex) VertexID() flow.Identifier { return v.collector.ResultID() }
 func (v *assignmentCollectorVertex) Level() uint64             { return v.collector.BlockHeight }
 func (v *assignmentCollectorVertex) Parent() (flow.Identifier, uint64) {
 	return v.collector.result.PreviousResultID, v.collector.BlockHeight - 1
 }
 
+func (v *assignmentCollectorVertex) SelectCollector() AssignmentCollector {
+	if v.processable {
+		return v.collector
+	} else {
+		return v.nonProcessableCollector
+	}
+}
+
+func (v *assignmentCollectorVertex) Update(processable bool) {
+	if processable != v.processable {
+		// if became unprocessable means it's orphan now
+		if v.processable {
+			v.nonProcessableCollector = NewOrphanAssignmentCollector(v.collector.ResultID(), v.collector.BlockID())
+		}
+
+		v.processable = processable
+	}
```

Comment:
Not sure how concurrency-safe this is:
* The `AssignmentCollectorTree` is locked while we update. 
* But I am wondering if other threads could hold a reference to the struct before its state transition?